src/Codec/Archive/Unpack.hs

```diff
@@ -96,28 +96,28 @@ unpackEntriesFp a fp = do
     case res of
         Nothing -> pure ()
         Just x  -> do
-            preFile <- liftIO $ archiveEntryPathname x
-            file <- liftIO $ peekCString preFile
-            let file' = fp </> file
-            liftIO $ withCString file' $ \fileC ->
-                archiveEntrySetPathname x fileC
-            ft <- liftIO $ archiveEntryFiletype x
-            case ft of
-                Just{} -> do
-                    ignore $ archiveReadExtract a x archiveExtractTime
-                    liftIO $ archiveEntrySetPathname x preFile
-                Nothing -> do
-                    preHardlink <- liftIO $ archiveEntryHardlink x
-                    hardlink <- liftIO $ peekCString preHardlink
-                    let hardlink' = fp </> hardlink
-                    liftIO $ withCString hardlink' $ \hl ->
-                        archiveEntrySetHardlink x hl
-                    ignore $ archiveReadExtract a x archiveExtractTime
-                    liftIO $ archiveEntrySetPathname x preFile
-                    liftIO $ archiveEntrySetHardlink x preHardlink
+            ignore . withPrefix fp x $ archiveReadExtract a x archiveExtractTime
             ignore $ archiveReadDataSkip a
             unpackEntriesFp a fp
 
+-- TODO: work with libarchive to get rid of this unsafe hack
+-- See https://github.com/libarchive/libarchive/issues/1203
+withPrefix :: FilePath -> ArchiveEntryPtr -> IO a -> IO a
+withPrefix fp x inner = do
+    file0 <- peekCString =<< archiveEntryPathname x
```

Comment:
Doesn't this mean the pointer to the original archive entry pathname is forgotten? Thus, the original pathname will never be freed.

Reply:
Declaration is `const char * archive_entry_pathname(struct archive_entry *a)`. Thus ownership for returned pointer is not transferred to caller  as in `char *strdup(const char*)`. It still owned by `archive_entry` structure.

You can find [similar code](https://github.com/vmchale/libarchive/blob/476b802d8b5ebf6dabb1b0a0c6b9789c2b64f0c3/src/Codec/Archive/Unpack.hs#L70) in `readEntry`:

```haskell
Entry
        <$> (peekCString =<< archiveEntryPathname entry)
        <*> readContents a entry
        -- ...
```